Source/JavaScriptCore/llint/WebAssembly64.asm

```diff
@@ -1257,3 +1257,31 @@ wasmAtomicCompareExchangeOps(_cmpxchg, Cmpxchg,
     end)
 end
 
+# GC ops
+
+wasmOp(i31_new, WasmI31New, macro(ctx)
+    mloadi(ctx, m_value, t0)
+    andq 0x7fffffff, t0
+    orq TagNumber, t0
+    returnq(ctx, t0)
+end)
+
+wasmOp(i31_get_s, WasmI31GetS, macro(ctx)
+    mloadp(ctx, m_ref, t0)
+    bqeq t0, ValueNull, .throw
+    lshifti 0x1, t0
+    rshifti 0x1, t0
+    returni(ctx, t0)
+
+.throw:
+    throwException(NullI31GetS)
+end)
+
+wasmOp(i31_get_u, WasmI31GetU, macro(ctx)
+    mloadp(ctx, m_ref, t0)
+    bqeq t0, ValueNull, .throw
+    returni(ctx, t0)
+
+.throw:
```

Comment:
> if i31 refs aren't nullable, then shouldn't downcasts and imports trap before getting here?

In general it should be possible for a non-null `i31ref` to get passed in here, they share a null with other reference types. For example the type `i31ref` is shorthand for `(ref null i31)` and includes null.

I think it might be possible in the B3/Air tiers to elide the null check though by inspecting the argument type and seeing if it allows null.